File: common/src/main/java/com/toddway/firebaserx/FirebaseRxCommon.java
package com.toddway.firebaserx;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import rx.Observable;
import rx.subscriptions.Subscriptions;

/**
 * Created by tway on 2/7/17.
 */

public class FirebaseRxCommon {

    public static Observable<DataSnapshot> observeValue(Query query) {
        return Observable.create(subscriber -> {
            final ValueEventListener listener = new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot snapshot) {
                    if (!subscriber.isUnsubscribed()) {
                        subscriber.onNext(snapshot);
                    }
                }

                @Override
                public void onCancelled(DatabaseError databaseError) {
                    int code = databaseError.getCode();
                    //if (code == -3) ; //permission denied, do nothing
                    subscriber.onError(databaseError.toException());
                }
            };

            subscriber.add(Subscriptions.create(() -> {
                query.removeEventListener(listener);
            }));

            query.addValueEventListener(listener);
        });
    }

    public static <T> Observable<T> observeValue(Query query, Class<T> type) {
        return observeValue(query).map(dataSnapshot -> dataSnapshot.getValue(type));
    }

    public static <T> Observable<Map<String, T>> observeChildMap(Query query, Class<T> type) {
        return observeValue(query)
                .flatMap(dataSnapshot -> toMap(dataSnapshot.getChildren(), type));
    }


    public static <T> Observable<List<T>> observeChildList(Query query, Class<T> type) {
        return observeValue(query)
                .flatMap(dataSnapshot -> toList(dataSnapshot.getChildren(), type));
    }

    private static <T> Observable<Map<String, T>> toMap(Iterable<DataSnapshot> dataSnapshots, Class<T> type) {
        return Observable.from(dataSnapshots)
                .toMap(
                        dataSnapshot1 -> dataSnapshot1.getKey(),
                        dataSnapshot1 -> dataSnapshot1.getValue(type)
                );
    }

    private static <T> Observable<List<T>> toList(Iterable<DataSnapshot> dataSnapshots, Class<T> type) {
        return Observable.from(dataSnapshots)
                .map(dataSnapshot2 -> dataSnapshot2.getValue(type))
                .toList();
    }

    public static <T> Observable<Map<String, T>> observeChildMap(Query query, Class<T> type, Iterable<String> keys) {
        return observeEqualTos(query, keys)
                .switchMap(snapshots -> toMap(snapshots, type));
    }

    private static Observable<List<DataSnapshot>> observeEqualTos(Query query, Iterable<String> equalTos) {
        List<Observable<DataSnapshot>> observables = new ArrayList<>();
        if (equalTos != null) {
            for (String equalTo : equalTos) {
                Observable<DataSnapshot> o = observeValue(query.equalTo(equalTo));
                observables.add(o);
            }
        }

        if (observables.isEmpty()) return Observable.just(new ArrayList<>());
        else return Observable
                .combineLatest(observables, args -> {
                    List<DataSnapshot> list = new ArrayList<>();
                    for (Object arg : args) list.add((DataSnapshot) arg);
                    return list;
                })
                .flatMap(dataSnapshots -> Observable
                        .from(dataSnapshots)
                        .flatMap(dataSnapshot -> Observable.from(dataSnapshot.getChildren()))
                        .toList()
                );
    }
}

File: common/src/main/java/com/toddway/firebaserx/FirebaseRxCommon.kt
package com.toddway.firebaserx

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.Query
import com.google.firebase.database.ValueEventListener
import rx.Observable
import rx.subscriptions.Subscriptions
import java.util.*

/**
 * Created by tway on 2/7/17.
 */


fun Query.observeValue(): Observable<DataSnapshot> {
    return Observable.create<DataSnapshot> { subscriber ->
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!subscriber.isUnsubscribed) {
                    subscriber.onNext(snapshot)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                val code = databaseError.code
                //if (code == -3) ; //permission denied, do nothing
                subscriber.onError(databaseError.toException())
            }
        }

        subscriber.add(Subscriptions.create { this.removeEventListener(listener) })

        this.addValueEventListener(listener)
    }
}

fun <T> Query.observeValue(type: Class<T>): Observable<T> {
    return observeValue().map { snap -> snap.getValue(type) }
}

fun <T> Query.observeChildMap(type: Class<T>): Observable<Map<String, T>> {
    return observeValue().map { toMap(it.children, type) }
}

fun <T> Query.observeChildMap(type: Class<T>, keys: Iterable<String>): Observable<Map<String, T>> {
    return observeEqualTos(this, keys).map { toMap(it, type) }
}

private fun <T> toMap(snaps: Iterable<DataSnapshot>, type: Class<T>): Map<String, T> {
    val map = LinkedHashMap<String, T>()
    snaps.forEach { snap ->
        snap.getValue(type)?.let { value ->
            map.put(snap.key, value)
        }
    }
    return map
}


private fun observeEqualTos(query: Query, equalTos: Iterable<String>?): Observable<List<DataSnapshot>> {
    val observables = ArrayList<Observable<DataSnapshot>>()
    if (equalTos != null) {
        for (equalTo in equalTos) {
            val o = query.equalTo(equalTo).observeValue()
            observables.add(o)
        }
    }

    if (observables.isEmpty())
        return Observable.just(ArrayList<DataSnapshot>())
    else
        return Observable
                .combineLatest(observables) { args ->
                    args.map { it as DataSnapshot }
                }
                .flatMap { dataSnapshots ->
                    Observable
                            .from(dataSnapshots)
                            .flatMap({ dataSnapshot : DataSnapshot -> Observable.from(dataSnapshot.children) })
                            .toList()
                }
}